onnxruntime/contrib_ops/cpu/bert/decoder_masked_multihead_attention.cc

```diff
@@ -0,0 +1,491 @@
+                                                                      scale_,
+                                                                      is_unidirectional,
+                                                                      past_present_share_buffer_,
+                                                                      kDecoderMaskedMultiHeadAttention));
+
+  int batch_size = parameters.batch_size;
+  int sequence_length = parameters.sequence_length;
+  int head_size = parameters.head_size;
+  int v_head_size = parameters.v_head_size;
+  int hidden_size = parameters.hidden_size;
+  int v_hidden_size = parameters.v_hidden_size;
+
+  // This kernel is for decoding only (i.e.) sequence length has to be 1
+  if (sequence_length != 1) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Input sequence length should be 1 to use DecoderMaskedMultiHeadAttention. "
+                           "Actual length is ",
+                           sequence_length);
+  }
+
+  if (head_size != v_head_size) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, NOT_IMPLEMENTED,
+                           "QK head size should be same as V head size to use DecoderMaskedMultiHeadAttention");
+  }
+
+  if (parameters.mask_type != AttentionMaskType::MASK_2D_KEY_PADDING &&
+      parameters.mask_type != AttentionMaskType::MASK_NONE) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, NOT_IMPLEMENTED,
+                           "DecoderMaskedMultiHeadAttention only supports no mask or 2D key "
+                           "padding mask of shape [batch, total_seq_length] currently");
+  }
+
+  TensorShapeVector output_shape(3);
+  output_shape[0] = static_cast<int64_t>(batch_size);
+  output_shape[1] = static_cast<int64_t>(sequence_length);
+  output_shape[2] = static_cast<int64_t>(v_hidden_size);
+  Tensor* output = context->Output(0, output_shape);
+
+  std::vector<int64_t> present_dims{
+      parameters.batch_size, parameters.num_heads,
+      past_present_share_buffer_ ? parameters.max_sequence_length : parameters.total_sequence_length,
+      head_size};
+  TensorShape present_shape(present_dims);
+  Tensor* present_key = context->Output(kPresentOutputIndex, present_shape);
+  Tensor* present_value = context->Output(kPresentOutputIndex + 1, present_shape);
+  Tensor* cross_qk = nullptr;
+
+  // Decoder cross-attention
+  if (past_key == nullptr && present_key == nullptr) {
+    if (attention_bias != nullptr) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, NOT_IMPLEMENTED,
+                             "DecoderMaskedMultiHeadAttention does not support attention bias for cross-attention");
+    }
+
+    parameters.is_cross_attention = true;
+    parameters.total_sequence_length = parameters.kv_sequence_length;
+    parameters.max_sequence_length = parameters.kv_sequence_length;
+    // parameters.k and parameters.v are nullptr
+    parameters.k_cache = const_cast<T*>(key->Data<T>());
```

Comment:
It's tricky to use k_cache instead of k to pass the key value. It is possible to pass by k, then use parameters.is_cross_attention==true to decide whether to access k or k_cache later.